crates/ruff_python_index/src/fstring_ranges.rs

```diff
@@ -89,7 +92,12 @@ impl FStringRangesBuilder {
     }
 
     pub(crate) fn finish(self) -> FStringRanges {
-        debug_assert!(self.start_locations.is_empty());
+        if !self.start_locations.is_empty() {
+            debug!(
+                "Unterminated f-strings detected at: {:?}",
+                self.start_locations
+            );
+        }
```

Comment:
I guess this isn't really required because we'd be providing diagnostics for the same.

Reply:
I wonnder if it would be useful to store the range with a missing and position instead of omitting it entirely.

Reply:
To extend on this:  It might even be more correct to simply assume that everything after the fstring start token is inside the fstring if the end token is missing. However, what the correct interpretation is might depend on the specific usage. E.g. adding noqa comments might be a bad idea for unclosed fstring. Rules testing if the range is part of a multiline string could get away by assuming the string goes to the end of the file

Reply:
I thought of the same in one of my [proposed solution](https://github.com/astral-sh/ruff/issues/8065#issuecomment-1776530001).

For rules, they're only being used to check `W605` (invalid escape sequence) and `ISC` (implicit string concatenation) which will be fine. For the former, the fix will be created (see below) while for later the fix won't be created.

```diff
-f"\.png{
+rf"\.png{
```

The NoQA logic will need to be looked into though. We might have to somehow mark these ranges as "incomplete from source code point of view".

Reply:
I'm also fine with adding the entire range, but just omitting the range of an unterminated string seems wrong to me (I rather overestimate than underestimate the f-string range)

Reply:
We should probably avoid adding NoQA directives then otherwise it won't be lexed as a comment and the `--add-noqa` will keep on adding them.

```python
f"\.png
  # noqa: W605
  # noqa: W605
  # ...
```

I'm not sure how to store this info. One solution could be to update the signature from `BTreeMap<TextSize, TextRange>` to `BTreeMap<TextSize, FStringRange>` where:
```rust
struct FStringRange {
	range: TextRange,
	// Is this f-string complete i.e., does it have both the start and end tokens?
	complete: bool
}
```

Or, while computing the NoQA mapping we can ignore these ranges by somehow checking if they're for a _complete_ f-string or not.

@charliermarsh Any thoughts here?

Reply:
Earlier I though that ignoring such ranges here would work:

https://github.com/astral-sh/ruff/blob/29fb86e3c831fd6e72283c0342f85e2b2fa7479c/crates/ruff_linter/src/directives.rs#L154-L171

But, I don't think so that will as then by default Ruff will add the directive at the end of the line.

I think we can update this function to return a `Option<TextSize>` instead:

https://github.com/astral-sh/ruff/blob/29fb86e3c831fd6e72283c0342f85e2b2fa7479c/crates/ruff_linter/src/noqa.rs#L746-L762

which will be used here and `continue` on `None`:

https://github.com/astral-sh/ruff/blob/29fb86e3c831fd6e72283c0342f85e2b2fa7479c/crates/ruff_linter/src/noqa.rs#L527-L527

Note that even if we somehow support this, it'll be impossible to ignore any violation on that line manually. For example,

```python
# Unterminated f-string
f"\.png{x}  # noqa: W605
```

Here, the `noqa` comment is part of the f-string and is not a comment.

Reply:
The plan you described here makes sense to me. Thanks for walking me through it.

Reply:
I've a local branch with a possible implementation for this but there are a few complexities involved. I'll proceed with merging this and open a new PR with that change instead.